(#52) Replacing postgresql-embedded with Testcontainers

scristalli opened the pull request:
This is for #52.

As I explained in [my comment](https://github.com/g4s8/ghman/issues/52#issuecomment-510864593), I think postgresql-embedded is broken, so I'm suggesting we switch to Testcontainers.

While this PR works locally (with a functional Docker environment available), I left a puzzle to adapt CircleCI, because Testcontainers needs `machine` instead of `docker` to work in CircleCI ([reference](https://www.testcontainers.org/supported_docker_environment/continuous_integration/circle_ci/)).

0crat commented:
Job [#70](https://github.com/g4s8/ghman/issues/70) is now in scope, role is `REV`

<!-- https://www.0crat.com/footprint/CJT4R32Q3/23aed309-2bc0-4ed1-8b04-56675e9b456d, version: 0.48.2, hash: ${buildNumber} -->

0crat commented:
This pull request [#70](https://github.com/g4s8/ghman/issues/70) is assigned to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), here is [why](https://www.0crat.com/footprint/CJT4R32Q3/599b2a28-ea6c-47df-b681-d3452b7c0932); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

<!-- https://www.0crat.com/footprint/CJT4R32Q3/e968c9ad-b39c-4a07-a64d-a04c06c6ea81, version: 0.48.2, hash: ${buildNumber} -->

olenagerasimova commented on src/test/java/com/g4s8/ghman/data/DatabaseExtension.java:
```diff
@@ -41,23 +43,41 @@
+                this.postgres.getPassword()
+            ).asString()
+        );
+        new JdbcSession(this.src).sql(
```
@scristalli what is wrong with `Flyway` migrations? Will we have to add here each new sql script manually?

scristalli replied:
@olenagerasimova you are absolutely right, major mistake of mine. Restored `Flyway` in commit 2e81d89ea35269f8cdc1f01c8e7b804811c762a8.

olenagerasimova commented on src/main/java/com/g4s8/ghman/App.java:
```diff
@@ -51,11 +50,7 @@ private App() {
     @SuppressWarnings("PMD.CyclomaticComplexity")
     public static void main(final String[] args) throws Exception {
         final Options opts = new Options(new IterableOf<>(args));
-        final DataSource data = new FlywayDataSource(
```
@scristalli how db schema will be set for the application? It will not work, if there is no structure for this application.

scristalli replied:
@olenagerasimova same here, restored in commit 2e81d89ea35269f8cdc1f01c8e7b804811c762a8.

olenagerasimova commented on src/test/java/com/g4s8/ghman/data/DatabaseExtension.java:
```diff
@@ -41,23 +43,41 @@
+                "&password=",
+                this.postgres.getPassword()
+            ).asString()
+        );
```
@scristalli btw will not it work if you set user and password by calling `setUser` and `setPassword` on `src` to avoid this joining? May you even will not need to specify username and password in `afterEach` when getting connection?

scristalli replied:
@olenagerasimova you're right, I was so busy [discussing the issue](https://github.com/g4s8/ghman/issues/52) of not having these methods on plain `DataSource` objects that I completely missed them in `PGSimpleDataSource`. Addressed in commit 60028f9.

scristalli commented:
@olenagerasimova thank you for the review and sorry for the mess; I overcomplicated this and ended up pushing unnecessary code, definitely not my best PR ever. I should have addressed all your comments.

olenagerasimova commented:
@scristalli thanks

olenagerasimova commented:
@rultor merge

rultor commented:
> @rultor merge

@olenagerasimova Thanks for your request. @g4s8 Please confirm this.

g4s8 commented:
@scristalli it's out of scope of #52 ticket, the ticket doesn't ask changing database integration test framework. Furthermore, it's not supposed to change CI environment and replace current docker container, and Rultor doesn't support sidecar docker containers.

0crat commented:
Job was finished in 15 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

<!-- https://www.0crat.com/footprint/CJT4R32Q3/56927298-7bdc-40d5-8b6a-8052c4bc97c1, version: 0.48.2, hash: ${buildNumber} -->

0crat commented:
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) please review this job completed by @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/CJT4R32Q3/46098981-f9ff-49cd-8438-5a5e472ff176, version: 0.48.2, hash: ${buildNumber} -->

0crat commented:
Job [#70](https://github.com/g4s8/ghman/issues/70) is not in the agenda of @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova), can't inspect

<!-- https://www.0crat.com/footprint/CJT4R32Q3/72823f97-c8ee-4675-a52d-b1f14b0e7480, version: 0.48.2, hash: ${buildNumber} -->

0crat commented:
The job [#70](https://github.com/g4s8/ghman/issues/70) is now out of scope

<!-- https://www.0crat.com/footprint/CJT4R32Q3/c9e41fb7-757c-491e-8c14-a922bfae7dba, version: 0.48.2, hash: ${buildNumber} -->

0crat commented:
Pull request [#70](https://github.com/g4s8/ghman/issues/70) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

<!-- https://www.0crat.com/footprint/CJT4R32Q3/8c1fdab5-b574-4cb5-a8c3-b87fda826c41, version: 0.48.2, hash: ${buildNumber} -->

ammaratef45 commented:
@0crat quality good

0crat commented:
Order was finished, quality is "good": +25 point(s) just awarded to @olenagerasimova[/z](https://www.0crat.com/u/olenagerasimova)

<!-- https://www.0crat.com/footprint/CJT4R32Q3/4ba8815d-c1f0-45d9-a909-9fa8431e84b3, version: 0.48.2, hash: ${buildNumber} -->

0crat commented:
Quality review completed: +4 point(s) just awarded to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45)

<!-- https://www.0crat.com/footprint/CJT4R32Q3/100b5e46-44ec-4819-ab88-6cbcdc6114b9, version: 0.48.2, hash: ${buildNumber} -->
